libraries/chain/include/steem/chain/index.hpp

```diff
@@ -58,3 +58,23 @@ void add_plugin_index( database& db )
 }
 
 } }
+
+#define STEEM_ADD_CORE_INDEX(db, index_name)                                                                 \
+   do {                                                                                                      \
+      steem::chain::add_core_index< index_name >( db );                                                      \
+      steem::chain::index_delegate delegate;                                                                 \
+      delegate.set_index_type =                                                                              \
+         []( database& _db, mira::index_type type, const boost::filesystem::path& p, const boost::any& cfg ) \
+            { _db.get_mutable_index< index_name >().mutable_indices().set_index_type( type, p, cfg ); };     \
+      db.set_index_delegate( #index_name, std::move( delegate ) );                                           \
+   } while( false );
```

Comment:
Why is this a do/while? Because of the repeated symbol name `delegate`?

First of all, you can encapsulate any arbitrary code block in braces without needed the do/while. Second, passing around `std::function` directly as lambdas allows you do set the index delegate without needing any in scope declarations and cleans up this macro a fair bit.

Reply:
The purpose of the do/while is to ensure correct behavior in conditionals. Consider the following:
```
#include <iostream>
#include <string>

#define MY_MACRO(x)                 \
    do {                            \
        std::cout << "multi\n";     \
        std::cout << "statement\n"; \
        std::cout << x << std::endl;\
    } while ( false )

#define MY_OTHER_MACRO(x)           \
    {                               \
        std::cout << "multi\n";     \
        std::cout << "statement\n"; \
        std::cout << x << std::endl;\
    }

int main()
{
  bool condition = true;
  
  if ( condition )
    MY_MACRO("test");
  else
    std::cout << "fail" << std::endl;

  return 0;
}
```

In this case `MY_MACRO` would behave as expected and `MY_OTHER_MACRO` would cause a compilation error, or worse yet erroneously cause the end of the if statement in other cases.

However, I did accidentally add `;` to the macro, and this should be removed to preserve this behavior in my opinion.

Reply:
I think this is a bit overkill because of the role of this macro being something internal.

I agree that if the macro were being used in a library, then creating it in such a manner would be preferred. As it stands, it is not incorrect.